src/v/iceberg/merge_append_action.cc

```diff
@@ -351,6 +351,160 @@ merge_append_action::upload_as_manifest(
+
+    template<typename TType>
+    primitive_value operator()(date_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, date_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(time_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, time_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(timestamp_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, timestamp_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(timestamptz_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, timestamptz_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(string_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, string_type>) {
+            return std::move(v);
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(uuid_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, uuid_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(fixed_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, fixed_type>) {
+            return std::move(v);
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(binary_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, binary_type>) {
+            return std::move(v);
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TType>
+    primitive_value operator()(decimal_value v, const TType& t) const {
+        if constexpr (std::is_same_v<TType, decimal_type>) {
+            return v;
+        }
+        throw_invalid(v, t);
+    }
+
+    template<typename TVal, typename TType>
+    [[noreturn]] void throw_invalid(const TVal& v, const TType& t) const {
+        throw std::invalid_argument(fmt::format(
+          "Cannot promote partition field value {} to type {}", v, t));
+    }
+};
+
+// Depending on how schema evolves, type of partition key fields can change
+// (e.g. a field type can be promoted from int to long). This means that
+// partition values of all data files in the new manifest need to be
+// promoted to the common type before creating the manifest.
+void promote_partition_key_type(
+  partition_key& pk, const partition_key_type& pk_type) {
+    vassert(
+      pk.val->fields.size() == pk_type.type.fields.size(),
+      "unexpected partition key size: {} (expected: {})",
+      pk.val->fields.size(),
+      pk_type.type.fields.size());
```

Comment:
Should we be defensive here and not crash but throw? If there is a bug in a catalog maybe it's possible the data file partition doesn't match the partition spec? I always get a little nervous asserting on invariants enforced by external systems.

Reply:
Yes good point, we check this for new files, but don't check for existing manifests.